gstreamer-pbutils/src/encoding_profile.rs

```diff
@@ -0,0 +1,578 @@
+            }
+
+            fn enabled(mut self, enabled: bool) -> $name<'a> {
+                self.base.enabled = enabled;
+                self
+            }
+        }
+    }
+);
+
+fn set_common_fields<T: EncodingProfileEdit>(profile: &mut T, base_data: &EncodingProfileBuilderData) {
+    profile.set_name(base_data.name);
+    profile.set_description(base_data.description);
+    profile.set_preset(base_data.preset);
+    profile.set_preset_name(base_data.preset_name);
+    profile.set_allow_dynamic_output(base_data.allow_dynamic_output);
+    profile.set_enabled(base_data.enabled);
+    profile.set_restriction(base_data.restriction);
+    profile.set_presence(base_data.presence);
+}
+
+pub struct EncodingAudioProfileBuilder<'a> {
+    base : EncodingProfileBuilderData<'a>
+}
+
+declare_encoding_profile_builder!(EncodingAudioProfileBuilder);
+
+impl<'a> EncodingAudioProfileBuilder<'a> {
+    pub fn new() -> Self {
+        EncodingAudioProfileBuilder {
+            base: EncodingProfileBuilderData::new(),
+        }
+    }
+
+    pub fn build(self) -> Result<EncodingProfile, EncodingProfileBuilderError> {
+        if self.base.format.is_none() {
+            return Err(EncodingProfileBuilderError);
+        }
+
+        let profile = EncodingAudioProfile::new(
+            self.base.format.unwrap(), self.base.preset,
+            self.base.restriction, self.base.presence);
+
+        let mut profile = profile.upcast();
+        set_common_fields(&mut profile, &self.base);
+        Ok(profile)
+    }
+}
+
+pub struct EncodingVideoProfileBuilder<'a> {
+    base : EncodingProfileBuilderData<'a>,
+    pass: u32,
+    variable_framerate: bool,
+}
+
+declare_encoding_profile_builder!(EncodingVideoProfileBuilder);
+
+impl<'a> EncodingVideoProfileBuilder<'a> {
+    pub fn new() -> Self {
+        EncodingVideoProfileBuilder {
+            base: EncodingProfileBuilderData::new(),
+            pass: 0,
+            variable_framerate: false
+        }
+    }
+
+    pub fn pass(mut self, pass: u32) -> Self {
+        self.pass = pass;
+        self
+    }
+
+    pub fn variable_framerate(mut self, variable_framerate: bool) -> Self {
+        self.variable_framerate = variable_framerate;
+        self
+    }
+
+    pub fn build(self) -> Result<EncodingProfile, EncodingProfileBuilderError> {
+        if self.base.format.is_none() {
+            return Err(EncodingProfileBuilderError);
+        }
+
+        let video_profile = EncodingVideoProfile::new(
+            self.base.format.unwrap(), self.base.preset, self.base.restriction, self.base.presence);
+
+        video_profile.set_pass(self.pass);
+        video_profile.set_variableframerate(self.variable_framerate);
+
+        let mut profile = video_profile.upcast();
+        set_common_fields(&mut profile, &self.base);
+        Ok(profile)
+    }
+}
+
+pub struct EncodingContainerProfileBuilder<'a> {
+    base : EncodingProfileBuilderData<'a>,
+    profiles: LinkedList<& 'a EncodingProfile>
```

Comment:
Better use a `Vec<_>`, `LinkedList` is almost never the right choice of data structure